### PR TITLE
Form stats increase each time the payment status changes from revoked to complete 

### LINKED ISSUE #722: Form stats increase each time the payment status changes from revoked to complete
As discussed with @mathetos, he confirmed that he was able to replicate same issue.

**Quick solution:**
As @mathetos suggested anybody who is facing this issue can use give tool to fix this issue.
Go to **Reports > Tools** and try the "recalculate" feature.

screen shot 2016-07-13 at 10 17 18 pm

### PR DESCRIPTION
**Pull request by ravinderk**
This will resolve #722

### CODE REVIEW

**Comment by DevinWalker** [comment]
I took some time to review this and here's what I'm thinking.

When a `complete` payment goes to `revoked` status that amount is *reduced* from the donation form and donor stats. Currently, it is not.

When a payment goes from `revoked` to `complete` status that amount is *added* to the attached donation form and donor stats

1. Rather than an additional conditional and variable here, what about removing the `'revoked' === $this->status` from the conditional here https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L642

2. Also remove from here https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L686
 
What do you think?

**Comment by ravinderk** [comment]
@DevinWalker I agree with you. We can remove `revoked` payment status from both if conditions.
But I have question

1.  Can we add separate function which decrease stats when user switch to `revoked`?
     For this we have to add `process_revoked` function here https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L1317

2. Do we have to remove `revoked` from these condition
https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L1480
https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L1515
https://github.com/WordImpress/Give/blob/release/1.6/includes/admin/payments/actions.php#L191
https://github.com/WordImpress/Give/blob/release/1.6/includes/admin/payments/actions.php#L243

**Comment by DevinWalker** [comment]
1. Yes, please

2. I do not believe you should remove it from: https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L1480 because this will cause a `revoked` -> `pending` transaction to be deducted, which is incorrect. 

Same with `revoked` -> `cancelled` https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L1515 the stats should not be deducted, so no change here

I'm pretty sure it needs to be removed here: https://github.com/WordImpress/Give/blob/release/1.6/includes/admin/payments/actions.php#L191 - note I did not do any testing on it

Same here: https://github.com/WordImpress/Give/blob/release/1.6/includes/admin/payments/actions.php#L243 - this looks like it should be removed. It's dealing with switching the donation form of a transaction. The amount of the `revoked` transaction should not be added or deducted when switching forms. The amount has already been deducted.

----

Basically, I would like `revoked` to work like `refunded`.

**Comment by ravinderk** [comment]
@DevinWalker 
2. I tested that if payment status change from `revoked->pending` (without remove `revoked` from if condition): https://github.com/WordImpress/Give/blob/release/1.6/includes/payments/class-give-payment.php#L1480 this `if` condition becomes false because our old status is `revoked` and then  cause a transaction to be deducted, which is incorrect.
This is same for  `revoked->cancelled` and `revoked->refund`

**Comment by DevinWalker** [comment]
Ravinder - I just tested this again and ran into an issue when going from `complete->revoked->complete` the amount of the form is not add *back* to the donation form when going back to the `complete` status. So, if I were to revoke the same payment again the donation amount keeps subtracting from the donation form incorrectly.

It should be `complete->revoked (amount subtracted + count reduced) -> complete (amount added + count increases)`.

**Comment by DevinWalker** [comment]
@ravinderk I think `$payments->revoked` also needs to be removed from `give_get_total_sales` here as well: https://github.com/WordImpress/Give/blob/master/includes/payments/functions.php#L772

**Comment by ravinderk** [comment]
@DevinWalker 
I found following code where we can delete `revoked` status. Can you check and confirm?

https://github.com/WordImpress/Give/blob/master/includes/admin/dashboard-widgets.php#L63#L66
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/class-export-earnings.php#L122
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-all-stats.php#L61
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-all-stats.php#L268
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-customer-stats.php#L107
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-form-stats.php#L65
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-form-stats.php#L150
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-income.php#L68
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-income.php#L120
https://github.com/WordImpress/Give/blob/master/includes/admin/reporting/tools/class-give-tools-recount-single-customer-stats.php#L81
https://github.com/WordImpress/Give/blob/master/includes/payments/class-give-payment.php#L685
https://github.com/WordImpress/Give/blob/master/includes/payments/functions.php#L630
https://github.com/WordImpress/Give/blob/master/includes/payments/functions.php#L687
https://github.com/WordImpress/Give/blob/master/includes/payments/functions.php#L772

**Comment by DevinWalker** [comment]
1. Yes
2. Yes
3. Yes
4. Yes
5. Yes
6. Yes
7. Yes
8. Yes
9. Yes
10. Yes
11. Yes
12. Yes
13. Yes
14.

**Comment by DevinWalker** [comment]
@ravinderk I notice an issue here after testing. When payments go from `revoked` -> `complete` the amount is not added back to the form total:

qp6k7qws4a

**Comment by DevinWalker** [comment]
My bad, I messed that testing up when switching to your branch. I just confirmed it's working nicely! Merging.
